File: backend/src/services/aiProvider.js
'use strict';
const axios = require('axios');
const {randomUUID} = require('crypto');
const {getGeminiCredentials, getGeminiApiKey, getGeminiModel, providerConfiguration} = require('../config/aiProviders');
const {groqContent, retryDelayMs} = require('./groqProvider');
const logger = require('../utils/logger');

function failureInfo(error, provider) {
  const status = Number(error.status || error.response?.status) || undefined;
  let reason = 'API_FAILURE';
  if (error.code === 'MISSING_API_KEY') reason = 'MISSING_API_KEY';
  else if ([401, 403].includes(status)) reason = 'INVALID_API_KEY_OR_ACCESS';
  else if (status === 429) reason = 'RATE_OR_QUOTA_LIMIT';
  else if (['ECONNABORTED', 'ETIMEDOUT'].includes(error.code) || error.name === 'APIConnectionTimeoutError') reason = 'TIMEOUT';
  else if (error.code === 'INVALID_AI_RESPONSE') reason = 'INVALID_RESPONSE';
  else if (error.code === 'UNSUPPORTED_CAPABILITY') reason = 'UNSUPPORTED_CAPABILITY';
  else if ([400, 404, 422].includes(status)) reason = 'MODEL_OR_REQUEST_CONFIGURATION';
  else if (['ECONNREFUSED', 'ENOTFOUND', 'EACCES'].includes(error.code) || error.name === 'APIConnectionError') reason = 'CONNECTION_FAILURE';
  return {provider, status, reason, ...(status === 429 && retryDelayMs(error) != null ? {retryAfterMs: retryDelayMs(error)} : {})}; // Never include raw SDK error messages, headers or request bodies.
}
function unavailable(failures) {
  const descriptions = {MISSING_API_KEY: 'key is missing', INVALID_API_KEY_OR_ACCESS: 'key or access was rejected', RATE_OR_QUOTA_LIMIT: 'rate limit or quota was reached', TIMEOUT: 'request timed out', INVALID_RESPONSE: 'response failed validation', MODEL_OR_REQUEST_CONFIGURATION: 'model or request configuration was rejected', UNSUPPORTED_CAPABILITY: 'required source retrieval is unavailable', CONNECTION_FAILURE: 'connection failed', API_FAILURE: 'API request failed'};
  const error = new Error(`Live AI is unavailable. ${failures.map(f => `${f.provider === 'groq' ? 'Groq' : f.credential === 'GEMINI_API_KEY2' ? 'Gemini key 2' : 'Gemini key 1'}: ${descriptions[f.reason]}`).join('; ')}. No AI content was saved. Please retry or check server AI configuration.`);
  return Object.assign(error, {status: 503, code: 'AI_PROVIDERS_UNAVAILABLE', failures});
}
function validSchema(value, schema, tolerant = false) {
  if (!schema) return true;
  const type = schema.type?.toLowerCase();
  if (type === 'object') return value != null && !Array.isArray(value) && typeof value === 'object' && (tolerant || (schema.required || []).every(key => Object.hasOwn(value, key))) && Object.entries(schema.properties || {}).every(([key, sub]) => !Object.hasOwn(value, key) || validSchema(value[key], sub, tolerant));
  if (type === 'array') return Array.isArray(value) && (schema.minItems == null || value.length >= schema.minItems) && (schema.maxItems == null || value.length <= schema.maxItems) && value.every(item => validSchema(item, schema.items, tolerant));
  if (type === 'integer' && !Number.isInteger(value) && !(tolerant && typeof value === 'string' && /^-?\d+$/.test(value.trim()))) return false;
  if (type === 'boolean' && typeof value !== 'boolean' && !(tolerant && (value === 'true' || value === 'false'))) return false;
  if (type === 'string' && typeof value !== 'string' && !(tolerant && (typeof value === 'number' || typeof value === 'boolean'))) return false;
  if (type === 'number' && (typeof value !== 'number' || !Number.isFinite(value)) && !(tolerant && typeof value === 'string' && /^-?\d+(\.\d+)?$/.test(value.trim()))) return false;
  if (type === 'integer' && Number.isInteger(value) && schema.enum && !schema.enum.includes(value)) return false;
  if (schema.enum && !(schema.enum.includes(value) || (tolerant && typeof value === 'string' && schema.enum.some(entry => typeof entry === 'string' && entry.toLowerCase() === value.toLowerCase())))) return false;
  const numeric = Number(value);
  if (schema.minimum != null && Number.isFinite(numeric) && numeric < schema.minimum) return false;
  if (schema.maximum != null && Number.isFinite(numeric) && numeric > schema.maximum) return false;
  return true;
}
function validateResponse(response, {json, schema, search, tolerant}) {
  const candidate = response.data?.candidates?.[0];
  const text = candidate?.content?.parts?.filter(part => !part.thought).map(part => part.text || '').join('').trim();
  const bad = () => Object.assign(new Error('Invalid provider response'), {code: 'INVALID_AI_RESPONSE'});
  if (candidate?.finishReason !== 'STOP' || !text) throw bad();
  if (json) {
    let value;
    try { value = JSON.parse(text.replace(/^```(?:json)?\s*/i, '').replace(/\s*```$/, '')); } catch { throw bad(); }
    if (!validSchema(value, schema, tolerant)) throw bad();
  }
  if (search && !response.sourceEvidence?.length && (!candidate.groundingMetadata?.groundingChunks?.length || !candidate.groundingMetadata?.groundingSupports?.length)) throw bad();
  return response;
}
async function geminiContent(options) {
  const apiKey = options.apiKey ?? getGeminiApiKey();
  if (!apiKey) throw Object.assign(new Error('Gemini key missing'), {code: 'MISSING_API_KEY'});
  const {prompt, system, json, schema, maxOutputTokens, timeout, search} = options;
  // Every Gemini feature uses this setting, including generation and review.
  const selected = getGeminiModel();
  // Gemini 2.5 Flash-Lite is unavailable to new users. Its replacement uses
  // thinking levels; retain the old budget format for explicit 2.5 overrides.
  const thinkingConfig = /^gemini-3[.-]/.test(selected)
    ? {thinkingLevel: /flash-lite/.test(selected) ? 'minimal' : 'low'}
    : {thinkingBudget: 0};
  const response = await axios.post(`https://generativelanguage.googleapis.com/v1beta/models/${encodeURIComponent(selected)}:generateContent`, {
    ...(system ? {systemInstruction: {parts: [{text: system}]}} : {}), contents: [{parts: [{text: prompt}]}],
    ...(search ? {tools: [{google_search: {}}]} : {}),
    generationConfig: {temperature: 0.15, maxOutputTokens, thinkingConfig, ...(json ? {responseMimeType: 'application/json'} : {}), ...(schema ? {responseSchema: schema} : {})},
  }, {timeout, headers: {'Content-Type': 'application/json', 'x-goog-api-key': apiKey}});
  return {...response, provider: 'gemini', data: {...response.data, provider: 'gemini'}};
}
function createAIProvider({gemini = geminiContent, groq = groqContent, log = logger} = {}) {
  return async function generateContent(options) {
    const requestId = randomUUID();
    const timeout = Math.max(1000, options.timeout || 30000);
    const deadline = Date.now() + timeout;
    const failures = [];
    const credentials = getGeminiCredentials();
    const geminiAttempts = credentials.length ? credentials : [{credential: 'GEMINI_API_KEY', apiKey: ''}];
    const providerList = [
      ...geminiAttempts.map(entry => ({provider: 'gemini', call: gemini, ...entry})),
      {provider: 'groq', call: groq},
    ];

    for (const [index, {provider, call, credential, apiKey}] of providerList.entries()) {
      const remaining = deadline - Date.now();
      const identity = {provider, ...(credential ? {credential} : {})};
      if (remaining <= 0) { failures.push({...identity, reason: 'TIMEOUT'}); continue; }
      // Both Gemini attempts share 40% of the deadline, leaving time for Groq.
      const budget = provider === 'gemini' ? Math.min(remaining, 20000, Math.max(1, Math.floor(timeout * 0.4 / geminiAttempts.length))) : remaining;
      try {
        log.info('[AIProvider] Attempt', {requestId, ...identity, feature: options.feature || 'quiz'});
        const result = validateResponse(await call({...options, ...(provider === 'gemini' ? {apiKey} : {}), timeout: budget, maxOutputTokens: options.maxOutputTokens || 1200}), options);
        result.provider = provider;
        if (credential) result.credential = credential;
        result.data.provider = provider;
        result.failover = failures;
        log.info('[AIProvider] Success', {requestId, ...identity, fallback: failures.length > 0, model: result.data.modelVersion, responseId: result.data.responseId});
        return result;
      } catch (error) {
        const failure = {...failureInfo(error, provider), ...(credential ? {credential} : {})};
        failures.push(failure);
        log.warn('[AIProvider] Failure', {requestId, ...failure, nextProvider: providerList[index + 1]?.provider || null});
      }
    }
    throw unavailable(failures);
  };
}
const generateContent = createAIProvider();

module.exports = {generateContent, createAIProvider, geminiContent, failureInfo, validSchema, providerConfiguration};

File: ai-service/services/ai_provider.py
"""Shared live text provider for the Python service; credentials never leave headers."""
import json
import logging
import os
import time
import uuid
import requests
from services.ai_config import has_key, get_gemini_credentials, get_gemini_model, thinking_config

log = logging.getLogger('ai-provider')

class AIProviderError(RuntimeError):
    def __init__(self, status_code=503, message='Both live AI providers are unavailable.', retries=0):
        self.status_code = status_code
        self.api_message = message
        self.retries = retries
        super().__init__(message)

def generate_content(prompt, *, gemini_key=None, model=None, response_json=True, temperature=0.2, timeout=60):
    deadline = time.monotonic() + timeout
    request_id = str(uuid.uuid4())
    failures = []
    credentials = get_gemini_credentials() or [('GEMINI_API_KEY', gemini_key)]
    attempts = [('gemini', name, key) for name, key in credentials] + [('groq', 'GROQ_API_KEY', os.getenv('GROQ_API_KEY'))]
    for index, (provider, credential, key) in enumerate(attempts):
        label = 'Gemini key 2' if credential == 'GEMINI_API_KEY2' else 'Gemini key 1' if provider == 'gemini' else 'Groq'
        if not has_key(key):
            failures.append(f'{label}: missing key')
            log.warning('AI failure request=%s provider=%s reason=MISSING_KEY', request_id, provider)
            continue
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            failures.append(f'{label}: timeout')
            continue
        selected = get_gemini_model() if provider == 'gemini' else os.getenv('GROQ_MODEL', 'openai/gpt-oss-120b')
        log.info('AI attempt request=%s provider=%s credential=%s model=%s', request_id, provider, credential, selected)
        try:
            if provider == 'gemini':
                payload = {'contents':[{'parts':[{'text':prompt}]}], 'generationConfig':{'temperature':temperature,'maxOutputTokens':8000,'thinkingConfig':thinking_config(selected)}}
                if response_json:
                    payload['generationConfig']['responseMimeType'] = 'application/json'
                response = requests.post(f'https://generativelanguage.googleapis.com/v1beta/models/{selected}:generateContent',headers={'x-goog-api-key':key.strip()},json=payload,timeout=min(remaining,20,timeout*.4/len(credentials)))
            else:
                payload = {'model':selected,'messages':[{'role':'system','content':'Answer accurately. Return only a valid JSON object.' if response_json else 'Give accurate, concise educational guidance.'},{'role':'user','content':prompt}], 'temperature':temperature,'max_completion_tokens':8000}
                if response_json:
                    payload['response_format']={'type':'json_object'}
                if selected.startswith('openai/gpt-oss-'):
                    payload['reasoning_effort']='low'
                response = requests.post('https://api.groq.com/openai/v1/chat/completions',headers={'Authorization':f'Bearer {key.strip()}'},json=payload,timeout=remaining)
            response.raise_for_status()
            data=response.json()
            if provider == 'gemini':
                candidate=data.get('candidates',[{}])[0]
                text=''.join(p.get('text','') for p in candidate.get('content',{}).get('parts',[]) if not p.get('thought'))
                finished=candidate.get('finishReason')=='STOP'
            else:
                candidate=data.get('choices',[{}])[0]
                text=candidate.get('message',{}).get('content','')
                finished=candidate.get('finish_reason')=='stop'
            if not finished or not isinstance(text,str) or not text.strip():
                raise ValueError('invalid response')
            if response_json:
                json.loads(text)
            log.info('AI success request=%s provider=%s credential=%s fallback=%s',request_id,provider,credential,bool(failures))
            return text.strip()
        except requests.Timeout:
            reason='timeout'
        except requests.HTTPError as error:
            status=error.response.status_code if error.response is not None else 0
            reason={401:'invalid key or access',403:'invalid key or access',429:'rate or quota limit'}.get(status,f'API failure ({status})')
        except requests.RequestException:
            reason='connection failure'
        except (ValueError,IndexError,KeyError,TypeError):
            reason='invalid response'
        failures.append(f'{label}: {reason}')
        log.warning('AI failure request=%s provider=%s credential=%s reason=%s next=%s',request_id,provider,credential,reason,attempts[index+1][0] if index+1<len(attempts) else 'none')
    raise AIProviderError(message='Live AI unavailable. ' + '; '.join(failures) + '. No generated content was saved.')
